File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/model/metrics/PerformanceMetrics.java
package com.study.webflux.rag.domain.model.metrics;

import java.time.Instant;
import java.util.List;
import java.util.Map;

public record PerformanceMetrics(
	String pipelineId,
	String status,
	Instant startedAt,
	Instant finishedAt,
	long totalDurationMillis,
	Long firstResponseLatencyMillis,
	Long lastResponseLatencyMillis,
	List<StagePerformance> stages,
	Map<String, Object> systemAttributes
) {
	public record StagePerformance(
		String stageName,
		String status,
		Instant startedAt,
		Instant finishedAt,
		long durationMillis,
		Map<String, Object> attributes) {
	}

	public static PerformanceMetrics fromPipelineSummary(
		String pipelineId,
		String status,
		Instant startedAt,
		Instant finishedAt,
		long durationMillis,
		Long firstResponseLatencyMillis,
		Long lastResponseLatencyMillis,
		List<StagePerformance> stages,
		Map<String, Object> systemAttributes) {
		return new PerformanceMetrics(
			pipelineId,
			status,
			startedAt,
			finishedAt,
			durationMillis,
			firstResponseLatencyMillis,
			lastResponseLatencyMillis,
			stages,
			systemAttributes);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/monitoring/model/PerformanceMetrics.kt
package com.miyou.app.domain.monitoring.model

import java.time.Instant

data class PerformanceMetrics(
    val pipelineId: String,
    val status: String,
    val startedAt: Instant?,
    val finishedAt: Instant?,
    val totalDurationMillis: Long,
    val firstResponseLatencyMillis: Long?,
    val lastResponseLatencyMillis: Long?,
    val stages: List<StagePerformance>,
    val systemAttributes: Map<String, Any?>,
) {
    data class StagePerformance(
        val stageName: String,
        val status: String,
        val startedAt: Instant?,
        val finishedAt: Instant?,
        val durationMillis: Long,
        val attributes: Map<String, Any?>,
    )

    companion object {
        fun fromPipelineSummary(
            pipelineId: String,
            status: String,
            startedAt: Instant?,
            finishedAt: Instant?,
            durationMillis: Long,
            firstResponseLatencyMillis: Long?,
            lastResponseLatencyMillis: Long?,
            stages: List<StagePerformance>,
            systemAttributes: Map<String, Any?>,
        ): PerformanceMetrics =
            PerformanceMetrics(
                pipelineId,
                status,
                startedAt,
                finishedAt,
                durationMillis,
                firstResponseLatencyMillis,
                lastResponseLatencyMillis,
                stages,
                systemAttributes,
            )
    }
}
